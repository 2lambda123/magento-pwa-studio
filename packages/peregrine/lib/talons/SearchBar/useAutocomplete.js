import { useEffect } from 'react';
import { useFieldState } from 'informed';

import { useQuery } from '../../hooks/useQuery';

/**
 * @typedef { import("graphql").DocumentNode } DocumentNode
 */

/**
 * Returns props necessary to render an Autocomplete component.
 * @param {Object} props
 * @param {DocumentNode} props.query - GraphQL query
 * @param {Boolean} props.visible - whether to show
 */
export const useAutocomplete = props => {
    const { query, visible } = props;

    // prepare to run query
    const [queryResult, queryApi] = useQuery(query);
    const { data, error, loading } = queryResult;
    const { resetState, runQuery, setLoading } = queryApi;

    // retrieve value from another field
    const { value } = useFieldState('search_query');
    const valid = value && value.length > 2;

    // determine message type
    const products = data && data.products;
    const hasResult = products && products.items;
    const resultCount = hasResult && products.items.length;
    let messageType = '';

    if (error) {
        messageType = 'ERROR';
    } else if (loading) {
        messageType = 'LOADING';
    } else if (!hasResult) {
        messageType = 'PROMPT';
    } else if (!resultCount) {
        messageType = 'EMPTY_RESULT';
    } else {
        messageType = 'RESULT_SUMMARY';
    }

    // run the query once on mount, and again whenever state changes
    useEffect(() => {
        if (visible && valid) {
            setLoading(true);
            runQuery({ variables: { inputText: value } });
        } else if (!value) {
            resetState();
        }
    }, [resetState, runQuery, setLoading, valid, value, visible]);

    return {
        hasResult,
        messageType,
        products,
        queryResult,
        resultCount,
        value
    };
};
