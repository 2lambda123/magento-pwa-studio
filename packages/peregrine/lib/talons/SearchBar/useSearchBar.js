import { useCallback, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { useDropdown } from '../../hooks/useDropdown';

const initialValues = { search_query: '' };

export const useSearchBar = () => {
    const [valid, setValid] = useState(false);
    const { elementRef, expanded, setExpanded } = useDropdown();
    const history = useHistory();
    const { push } = history;

    // expand or collapse on input change
    const handleChange = useCallback(
        value => {
            const hasValue = !!value;
            const isValid = hasValue && value.length > 2;

            setValid(isValid);
            setExpanded(hasValue);
        },
        [setExpanded, setValid]
    );

    // expand on focus
    const handleFocus = useCallback(() => {
        setExpanded(true);
    }, [setExpanded]);

    // navigate on submit
    const handleSubmit = useCallback(
        ({ search_query }) => {
            if (search_query != null && search_query.trim().length > 0) {
                push(`/search.html?query=${search_query}`);
                setExpanded(false);
            }
        },
        [push, setExpanded]
    );

    return {
        containerRef: elementRef,
        expanded,
        handleChange,
        handleFocus,
        handleSubmit,
        initialValues,
        setExpanded,
        setValid,
        valid
    };
};
