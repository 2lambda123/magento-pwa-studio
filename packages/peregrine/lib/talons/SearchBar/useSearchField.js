import { useCallback, useEffect, useRef } from 'react';
import { useFieldState, useFormApi } from 'informed';

import { useAppContext } from '../../context/app';
/**
 * Returns props necessary to render a SearchField component.
 */
export const useSearchField = () => {
    const [{ searchOpen }] = useAppContext();
    const inputRef = useRef();

    const { value } = useFieldState('search_query');
    const formApi = useFormApi();

    const resetForm = useCallback(() => {
        formApi.reset();
    }, [formApi]);

    // When the search field is opened focus on the input.
    useEffect(() => {
        if (searchOpen && inputRef.current) {
            inputRef.current.focus();
        }
    }, [searchOpen]);

    return {
        inputRef,
        resetForm,
        value
    };
};
