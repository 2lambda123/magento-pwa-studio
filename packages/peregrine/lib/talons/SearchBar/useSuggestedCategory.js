import { useCallback } from 'react';
import { useHistory, useLocation } from 'react-router-dom';
import useInternalLink from '../../hooks/useInternalLink';

import { DELIMITER } from '../FilterModal/helpers';

// TODO: derive from store config when available
const setSearchParams = (existing, options) => {
    const params = new URLSearchParams(existing);
    const { categoryId, label, searchValue } = options;

    params.set('query', searchValue);
    params.set('category_id[filter]', `${label}${DELIMITER}${categoryId}`);

    return `${params}`;
};

/**
 * Return props necessary to render a SuggestedCategory component.
 *
 * @param {Object} props
 * @param {String} props.categoryId - category
 * @param {Function} props.onNavigate - callback to fire on link click
 * @param {String} props.searchValue - search term
 */
export const useSuggestedCategory = props => {
    const { onNavigate, ...restProps } = props;
    const { createHref } = useHistory();
    const { search } = useLocation();

    const nextSearchParams = setSearchParams(search, restProps);
    const destination = createHref({
        pathname: '/search.html',
        search: nextSearchParams
    });

    const { handleClick: setLoader } = useInternalLink('category');

    const handleClick = useCallback(() => {
        if (typeof onNavigate === 'function') {
            onNavigate();
        }
        // Sets next root component to show proper loading effect
        setLoader();
    }, [onNavigate, setLoader]);

    return {
        destination,
        handleClick
    };
};
