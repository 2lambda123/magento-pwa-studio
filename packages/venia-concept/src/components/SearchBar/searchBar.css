.root {
    align-items: center;
    display: none;
    justify-content: center;
}

.root_open {
    composes: root;
    display: flex;
}

.form {
    align-items: center;
    display: grid;
    justify-items: stretch;
    margin-bottom: 0.5rem;
    width: 22.5rem;
}
