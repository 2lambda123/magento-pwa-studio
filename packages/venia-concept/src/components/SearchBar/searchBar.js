import React, { Component } from 'react';
import { bool, func, object, shape, string } from 'prop-types';
import { Form } from 'informed';

import getQueryParameterValue from '../../util/getQueryParameterValue';
import { SEARCH_QUERY_PARAMETER } from '../../RootComponents/Search/consts';
import Icon from 'src/components/Icon';
import TextInput from 'src/components/TextInput';
import Trigger from 'src/components/Trigger';

import classify from 'src/classify';
import defaultClasses from './searchBar.css';

const initialValues = {
    search_query: ''
};

const clearIcon = <Icon name="x" attrs={{ height: 18, width: 18 }} />;
const searchIcon = <Icon name="search" attrs={{ height: 18, width: 18 }} />;

// TODO: remove export here (update story and test)
export class SearchBar extends Component {
    static propTypes = {
        classes: shape({
            form: string,
            root: string,
            root_open: string
        }),
        executeSearch: func.isRequired,
        history: object.isRequired,
        location: object.isRequired,
        isOpen: bool
    };

    state = {
        dirty: false
    };

    componentDidMount() {
        const searchValueFromQueryString = getQueryParameterValue({
            location: this.props.location,
            queryParameter: SEARCH_QUERY_PARAMETER
        });

        this.formApi.setValue('search_query', searchValueFromQueryString);
    }

    get resetButton() {
        const { resetForm, state } = this;

        return state.dirty ? (
            <Trigger action={resetForm}>{clearIcon}</Trigger>
        ) : null;
    }

    handleChange = ({ values }) => {
        const dirty = !!values.search_query;
        this.setState({ dirty });
    };

    handleSubmit = ({ search_query }) => {
        this.props.executeSearch(search_query, this.props.history);
    };

    resetForm = () => {
        this.formApi.reset();
    };

    setApi = formApi => {
        this.formApi = formApi;
    };

    render() {
        const { props, resetButton } = this;
        const { classes, isOpen } = props;
        const className = isOpen ? classes.root_open : classes.root;

        return (
            <div className={className}>
                <Form
                    className={classes.form}
                    getApi={this.setApi}
                    initialValues={initialValues}
                    onChange={this.handleChange}
                    onSubmit={this.handleSubmit}
                >
                    <TextInput
                        field="search_query"
                        after={resetButton}
                        before={searchIcon}
                    />
                </Form>
            </div>
        );
    }
}

export default classify(defaultClasses)(SearchBar);
