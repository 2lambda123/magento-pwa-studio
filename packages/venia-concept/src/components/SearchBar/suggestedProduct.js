import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Price } from '@magento/peregrine';
import classify from 'src/classify';
import { Link, resourceUrl } from 'src/drivers';

import defaultClasses from './suggestedProduct.css';

const productUrlSuffix = '.html';

class SuggestedProduct extends Component {
    static propTypes = {
        handleOnProductOpen: PropTypes.func.isRequired,
        url_key: PropTypes.string.isRequired,
        small_image: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
        price: PropTypes.shape({
            regularPrice: PropTypes.shape({
                amount: PropTypes.shape({
                    currency: PropTypes.string,
                    value: PropTypes.number
                })
            })
        }).isRequired,
        classes: PropTypes.shape({
            root: PropTypes.string,
            productName: PropTypes.string,
            productImage: PropTypes.string
        })
    };

    render() {
        const {
            handleOnProductOpen,
            classes,
            url_key,
            small_image,
            name,
            price
        } = this.props;

        const productLink = resourceUrl(`/${url_key}${productUrlSuffix}`);

        return (
            <li className={classes.root}>
                <Link onClick={handleOnProductOpen} to={productLink}>
                    <img
                        className={classes.productImage}
                        alt={name}
                        src={resourceUrl(small_image, {
                            type: 'image-product',
                            width: 60
                        })}
                    />
                </Link>
                <Link
                    onClick={handleOnProductOpen}
                    className={classes.productName}
                    to={productLink}
                >
                    {name}
                </Link>
                <Link onClick={handleOnProductOpen} to={productLink}>
                    <Price
                        currencyCode={price.regularPrice.amount.currency}
                        value={price.regularPrice.amount.value}
                    />
                </Link>
            </li>
        );
    }
}

export default classify(defaultClasses)(SuggestedProduct);
