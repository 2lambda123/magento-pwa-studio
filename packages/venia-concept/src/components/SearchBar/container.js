import { connect } from 'react-redux';

import { executeSearch } from 'src/actions/app';
import SearchBar from './searchBar';

const mapDispatchToProps = { executeSearch };

export default connect(
    null,
    mapDispatchToProps
)(SearchBar);
