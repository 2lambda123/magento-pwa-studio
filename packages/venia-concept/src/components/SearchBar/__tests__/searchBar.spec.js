import React from 'react';
import TestRenderer from 'react-test-renderer';

import SearchBar from '../searchBar';

const buttonTypes = el => el.type === 'button';
const formTypes = el => el.type === 'form';
const inputTypes = el => el.type === 'input';

const executeSearchMock = jest.fn();
const props = {
    executeSearch: executeSearchMock,
    history: {},
    location: { search: '?query=test' },
    isOpen: true
};

afterEach(() => {
    executeSearchMock.mockReset();
});

test('renders the correct tree', () => {
    const tree = TestRenderer.create(<SearchBar {...props} />).toJSON();

    expect(tree).toMatchSnapshot();
});

test('the input field is seeded from the location', () => {
    const expected = 'test';

    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    const input = instance.find(inputTypes);

    expect(input.props.value).toBe(expected);
});

test('the reset button is visible when the input field is not empty', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    const buttons = instance.findAll(buttonTypes);

    expect(buttons).toHaveLength(1);
});

test('the reset button is not visible when the input field is empty', () => {
    // Force the input field to be empty.
    const testProps = {
        ...props,
        location: { search: '' }
    };

    const renderer = TestRenderer.create(<SearchBar {...testProps} />);
    const instance = renderer.root;

    const buttons = instance.findAll(buttonTypes);

    expect(buttons).toHaveLength(0);
});

test('entering text in the input causes the reset button to appear', () => {
    // Start with an empty input field.
    const testProps = {
        ...props,
        location: { search: '' }
    };

    const renderer = TestRenderer.create(<SearchBar {...testProps} />);
    const instance = renderer.root;

    // Confirm that the button is not present.
    const preChangeButtons = instance.findAll(buttonTypes);
    expect(preChangeButtons).toHaveLength(0);

    // Simulate entering text in the input field.
    const input = instance.find(inputTypes);
    input.props.onChange({
        target: { value: 'some text' }
    });

    // Test that the button appears.
    const postChangeButtons = instance.findAll(buttonTypes);
    expect(postChangeButtons).toHaveLength(1);
});

test('the reset button clears the input', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    // Test that there is some text in the input to start.
    const input = instance.find(inputTypes);
    expect(input.props.value).toBe('test');

    // Simulate clicking the reset button.
    const button = instance.find(buttonTypes);
    button.props.onClick();

    // Test that the input has been cleared.
    expect(input.props.value).toBe('');
});

// TODO: test fails because mock isn't called ... but it is?
test.skip('submitting the form executes the search', () => {
    const renderer = TestRenderer.create(<SearchBar {...props} />);
    const instance = renderer.root;

    // Simulate form submit.
    const form = instance.find(formTypes);
    form.props.onSubmit();

    // Test that executeSearch was called.
    expect(executeSearchMock).toHaveBeenCalledTimes(1);
    expect(executeSearchMock).toHaveBeenNthCalledWith(1, 'test', props.history);
});
