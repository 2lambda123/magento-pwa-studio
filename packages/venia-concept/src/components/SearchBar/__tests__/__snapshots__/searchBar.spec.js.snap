// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the correct tree 1`] = `
<div>
  <form
    onReset={[Function]}
    onSubmit={[Function]}
  >
    <span
      style={
        Object {
          "--iconsAfter": 1,
          "--iconsBefore": 1,
        }
      }
    >
      <span>
        <input
          name="search_query"
          onBlur={[Function]}
          onChange={[Function]}
          value="test"
        />
      </span>
      <span>
        <span
          dangerouslySetInnerHTML={
            Object {
              "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-search\\"><circle cx=\\"11\\" cy=\\"11\\" r=\\"8\\"></circle><line x1=\\"21\\" y1=\\"21\\" x2=\\"16.65\\" y2=\\"16.65\\"></line></svg>",
            }
          }
        />
      </span>
      <span>
        <button
          onClick={[Function]}
          type="button"
        >
          <span
            dangerouslySetInnerHTML={
              Object {
                "__html": "<svg xmlns=\\"http://www.w3.org/2000/svg\\" width=\\"18\\" height=\\"18\\" viewBox=\\"0 0 24 24\\" fill=\\"none\\" stroke=\\"currentColor\\" stroke-width=\\"2\\" stroke-linecap=\\"round\\" stroke-linejoin=\\"round\\" class=\\"feather feather-x\\"><line x1=\\"18\\" y1=\\"6\\" x2=\\"6\\" y2=\\"18\\"></line><line x1=\\"6\\" y1=\\"6\\" x2=\\"18\\" y2=\\"18\\"></line></svg>",
              }
            }
          />
        </button>
      </span>
    </span>
    <p />
  </form>
</div>
`;
